Restrict autoloaded paths to loaded apps (and other enhancements)

RobinMcCorkell opened the pull request:
Unloaded apps (aka not enabled apps) will not get any files autoloaded anymore, which can prevent an unloaded app becoming an attack vector if it is not regularly updated.

In addition, the autoloader will properly resolve symlinked app directories, to allow for a semi-common usecase where apps are stored elsewhere but are symlinked into the ownCloud directory. Fixes issues as noted in https://github.com/owncloud/core/pull/18396#issuecomment-137419945

@fossxplorer want to test this out with your symlinked apps?

cc @icewind1991 @MorrisJobke @PVince81 @LukasReschke

Commit:
Only add valid root for enabled apps

Commit:
Resolve autoloader valid roots before checking

Allows symlinked app directories to work properly

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/d7483d4b-3149-4901-a247-2d9a7150c2e2) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/f74a6d4b-66bd-42d8-85e2-f00834b4fe34) was created.

LukasReschke commented:
That's cool :+1:

Xenopathic commented:
Oh, I've inadvertently fixed some other bugs too :rocket: 

Fixes #18224, fixes #18317, fixes #18836, fixes #18305, fixes #18296, fixes #18322, fixes #18329, fixes #18333, ........ (yes, I included the issues that were marked as duplicates of the original one, simply to show the impact of the bug).

The line that fixes all those bugs is the move of `self::$loadedApps[] = $app;` from `loadApps()` to `loadApp()`. Give me a second to prepare backport PRs, assuming that's wanted @karlitschek ?

karlitschek commented:
nice :+1: please backport

icewind1991 commented:
Looks goods :+1:

Xenopathic commented:
CI hasn't run the full autotest suite, so I'm not merging just yet...

Commit:
Unique exception for invalid autoload paths, better handling

Background jobs are tolerant of stale entries left by disabled apps,
which will cause an autoload exception.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/dfed7f30-cb71-44ea-8431-7fafd04ac367) was created.

Xenopathic commented:
The latest commit makes the autoloader throw a unique exception class, which the background job runner catches to prevent stale jobs breaking things.

MorrisJobke commented:
> The line that fixes all those bugs is the move of self::$loadedApps[] = $app; from loadApps() to loadApp(). Give me a second to prepare backport PRs, assuming that's wanted @karlitschek ?

Correct. Because loadApp is also called during app upgrade :see_no_evil:

MorrisJobke commented:
Code looks good, fixes my issue and CI runs fine locally :+1:

oparoz commented:
Completely breaks apps which don't rely solely on unit tests...

What is the workaround for tests this time?!?

oparoz commented:
Traits are not loaded:

> [OCP\AutoloadNotAllowedException]
  Autoload path not allowed: /apps/gallery/controller/pathmanipulation.php

Xenopathic commented:
@oparos is your app being loaded in your tests? That is the requirement - that any app is properly loaded through `OC_App::loadApp()` to enable the autoload path.

Alternatively, add `\OC::$loader->addValidRoot('your/app/path/here');` to your bootstrap :smile:

oparoz commented:
@Xenopathic - Last time I tried, loading the app did not work. At least it didn't bring routes up, but I'll try again to see if it solves this one.

Xenopathic commented:
@nickvergessen when you see this on Monday, before you get too excited and revert immediately, I *will* be sending out a mailing list email to clarify the updated requirements, as soon as we find an appropriate solution with @oparoz

oparoz commented:
I can confirm that loading the app in the boostrap still does nothing. I suspect a bug in the loader, unless someone can show me an app which loads (1. load the app 2. check the navigation menu).

But adding this to the boostrap fixes the issue: `OC::$loader->addValidRoot('../');`

It just doesn't make sense to have to whitelist the app being tested, but it's better than nothing.

oparoz commented:
Apart from the problems, I do applaud the move to protect the system from dormant apps :clap:

Xenopathic commented:
@oparoz Well, for unit tests that line will work just fine, but you realise it doesn't work as expected? All paths are relative to the OC::$SERVERROOT, so you basically just disabled the autoloader verification for everything :laughing:

oparoz commented:
Arf, yes... Missing `__DIR__`

oparoz commented:
LoadApp() works, but there is a problem with loadApps() https://github.com/owncloud/core/issues/18863

nickvergessen commented:
> @nickvergessen when you see this on Monday, before you get too excited and revert immediately, I will be sending out a mailing list email to clarify the updated requirements, as soon as we find an appropriate solution with @oparoz

Sorry @Xenopathic but this is :hankey: 
Leaving developers with a known broken system. Please notify people in advance. This really makes developing an app a pain. Third breakage in 5 days on the same part. **Please take more time on your fixes and test them, before merging them.** All apps are publicly available on github. Check them out and run their unit tests locally, so you see what's the problem and stop messing it up.

So the now required code for the bootstrap.php is:
```php
\OC::$loader->addValidRoot(OC::$SERVERROOT . '/tests');
\OC_App::loadApp('<app name here>');
```

Xenopathic commented:
@nickvergessen Actually, I was wondering why we only load a minimal set of apps in bootstrap.php: https://github.com/owncloud/core/blob/master/tests/bootstrap.php#L14-L15. We should be loading all enabled apps, which would prevent these issues?

nickvergessen commented:
I'm not using that bootstrap file....
I don't want integration tests where all apps are registered and add there stuff:
https://github.com/owncloud/activity/blob/master/tests/bootstrap.php
https://github.com/owncloud/mail/blob/master/tests/bootstrap.php
https://github.com/owncloud/notifications/blob/master/tests/bootstrap.php
...

Xenopathic commented:
OK, so the consensus is that we should explicitly require that the app (and any dependencies of that app) are loaded with `OC_App::loadApp()` before unit tests are run?

oparoz commented:
@Xenopathic - The boostrap still works when I remove `\OC::$loader->addValidRoot(OC::$SERVERROOT . '/tests');`. Is that intended?

MorrisJobke commented:
> @Xenopathic - The boostrap still works when I remove \OC::$loader->addValidRoot(OC::$SERVERROOT . '/tests');. Is that intended?

Yep. The root is added on appLoad()

oparoz commented:
@MorrisJobke - OK, there may be a bug then:
https://github.com/owncloud/mail/pull/1058#issuecomment-138269601

oparoz commented:
@MorrisJobke - In Gallery, `addValidRoot` is also still required, even with loadApps();
Keep that in mind for the dev documentation.
